### ISSUE TITLE
Member class typedefs and weird ast content

### ISSUE THREAD

**Issue by Manu343726** [foonathan/cppast#58]
cppast version: https://github.com/Manu343726/cppast/commit/b58405bffb6b2a1d67d31cd22102a3a273ab85a9
libclang parser
libclang version: 4.0.0
Explanation of the error:

I'm trying to parse and examine the contents of classes declared as:

``` cpp
typdef struct
{
    int a, b, c;
} Foo;
```

I've written a function that tries to extract the cpp_class entity out of a cpp_type_alias, trying to match my requirements (anonymous struct type alias, struct declared in that same header, etc). If the type alias entity matches, getting the aliased class would be equivalent to:

``` cpp
type_safe::optional_ref<const cppast::cpp_class> class_typedef(index, typedef_entity)
{
    return type_safe::cref(typedef_entity.underlying_type().entity().get(index)[0]);
}
```

Things get interesting once I put the typedef struct as a class member (Which is my target use case):

``` cpp
struct Class
{
    typedef struct
    {
        int a, b;
    } InnerTypedefClass;

    int a, b;
};
```
My current AST "scanner" currently handles classes and enums only, so a typedef entity like `TypedefClass` there is ignored.

If I introduce a case for typedefs as members and apply the typedef struct matching logic described earlier, I successfully find a class declaration node. **But also a class declaration node is automagically added as a member of the `Class` class**.

My class members scanning system us just a plain call to visit on the class node (See https://gitlab.com/Manu343726/tinyrefl/blob/master/tool/tool.cpp#L267), and it's from inside that visitor that the typedef check algorithm is run (An extra case for the switch for the typedef):

``` cpp
case cppast::cpp_entity_kind::cpp_type_alias:
    const auto alias = class_typedef(child);
    if(alias) std::cout << " - class typedef " << child.name() << "\n";
    break;

...
```

With that change only (Note there's nothing done with the alias in the switch side), I see a `TypedefClass` class declaration appearing as class declaration member of `Class`, as if already were a child of the class. Note also I get the right class name (from the typedef pov, not an empty class name). Tool output:

```
    - (class typedef) TypedefStruct
    - (class) TypedefStruct (declaration) [attributes: ] <-- MAGIC
```

That's fairly weird, but maybe I'm misunderstanding something about how the AST works.

After that, I said "Ok, if a child is automatically added to the class, let's take advantage of that".  Now I register typedef classes that are automatically added to my parent class to be scanned later. When I run an scan on `TypedefStruct` cpp_class entity, I get **no members, and parent class members as base classes of `TypedefStruct`**:

```
 # my_namespace::MyClass [attributes: ]
    - (member) f [attributes: ]
    - (member) str [attributes: ]
    - (enum) Enum [attributes: ]
    - (class) Foo (definition) [attributes: ]
    - (class) InnerClassWithMembers (definition) [attributes: ]
[warning] Found class with empty name at my_namespace::MyClass
    - (class typedef) TypedefStruct
    - (class) TypedefStruct (declaration) [attributes: ]
    - (member) innerClassInstance [attributes: ]
    - (base) my_namespace::BaseClass
    - (base) my_namespace::Foo
 # my_namespace::MyClass::TypedefStruct [attributes: ]
    - (base) my_namespace::MyClass::Foo                                  // WTF ???
    - (base) my_namespace::MyClass::InnerClassWithMembers // WTF ???
    - (base) my_namespace::MyClass::                                        // WTF ??? 
    - (base) my_namespace::MyClass::TypedefStruct                  // WHAAAAT ???
    - (base) my_namespace::MyClass::innerClassInstance         // WTF ???
    - (base) my_namespace::MyClass::private                             // WTF ???
    - (base) my_namespace::MyClass::_private                           // WTF ???
```

base classes are retrieved with `cpp_class::bases()`.

I will try to get a minimal repo ASAP.

**Comment by Manu343726** [comment]
BTW my only interest with all this is to correctly reflect classes declared as typedefs, so if you've a better idea for that I will be glad to hear about it.

**Comment by foonathan** [comment]
When I run `./cppast file.cpp` on

```cpp
struct Class
{
    typedef struct
    {
        int a, b;
    } InnerTypedefClass;

    int a, b;
};
```

I get:

```
AST for 'file.cpp':
+-Class (class) [definition]: `struct Class;`
  |-<anonymous> (class) [definition]: `struct ;`
  | |-a (member variable): `int a;`
  | +-b (member variable): `int b;`
  |-InnerTypedefClass (type alias): `using InnerTypedefClass=InnerTypedefClass;`
  |-a (member variable): `int a;`
  +-b (member variable): `int b;`
```

Which seems correct.

**Comment by Manu343726** [comment]
One quick question: Why the compilation of the tool is restricted to the case where cppast is the root cmake project?

**Comment by foonathan** [comment]
Isn't there an option to turn it one manually?

**Comment by Manu343726** [comment]
Yes, with the `CPPAST_BUILD_TOOL` variable, but the if looks like this:

``` cmake
if(${CPPAST_BUILD_TOOL} AND (CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR))
```

which disables the compilation if I'm using cppast as a dependency of an upper cmake project

**Comment by Manu343726** [comment]
I just was curious about what was behind that condition

**Comment by foonathan** [comment]
oh, that's supposed to be an OR

**Comment by Manu343726** [comment]
Fixed. Now I can go back to the original issue xD

**Comment by Manu343726** [comment]
I'm building a buildroot distro from scratch, so we have aprox 2hj to talk xD

```
AST for 'file.cpp':
+-Class (class) [definition]: `struct Class;`
  |-<anonymous> (class) [definition]: `struct ;`
  | |-a (member variable): `int a;`
  | +-b (member variable): `int b;`
  |-InnerTypedefClass (type alias): `using InnerTypedefClass=InnerTypedefClass;`
  |-a (member variable): `int a;`
  +-b (member variable): `int b;`
```

I see that the typedef points to itself, there's no clear relation in the AST that InnerTypedefClass is aliasing the anonymous class above. Do you think that it's safe to assume the typedef would always be the next child?

**Comment by foonathan** [comment]
>  see that the typedef points to itself, there's no clear relation in the AST that InnerTypedefClass is aliasing the anonymous class above.

Yes, this might be an issue.

> Do you think that it's safe to assume the typedef would always be the next child?

What do you mean by that?

**Comment by Manu343726** [comment]
> What do you mean by that?

that in the current AST walking scheme, a `typedef struct {} Foo` will always be reflected as a first node pointing to an anonymous struct, then a node pointing to a typedef

**Comment by foonathan** [comment]
What is the status of this problem?

**Comment by Manu343726** [comment]
I haven't found a solution. Once I merge the Jinja2 based tool of tinyrefl (I have rewritten the AST visitation code from scratch) I will check the issue again. 

Anyway, one possible solution to this kind of issue is allowing users to inherit cpp_entity/type and write their custom AST nodes, and injecting those if some properties are met. For example, if  cppast supported matchers (it's still in my to-do list) one would write something like this:

class StructTypedef : cpp_class { ... };

class StructTypedef storm : AstTransform<StructTypedef>
{
    bool matches(node) override
    {
        return TypedefStructMatcher.matches(node);
    }

    TypedefStruct replace(node) override
    {
        // replace node by virtual TypedefStruct node
    }
};

parser.addTransform(TypedefStructTransform);


*Sorry for the format, I'm on mobile*

TL;DR do you think that could cppast potentially support AST transformations like libtooling?

**Comment by foonathan** [comment]
> TL;DR do you think that cppast could potentially support AST transformations like libtooling?

If someone goes through the trouble of implementing it, yes, it can.

**Comment by foonathan** [comment]
So what is the actual status of this issue now?

**Comment by Manu343726** [comment]
I still cannot parse anonymous structs like that, but since I'm rewriting my parser completely I will close this so you get less noise in your issue board. When I'm done with the refactoring I will try to repro this issue again.
